Modify esmeralda

mmkekic opened the pull request:
This PR adds no-filter KDST/MC/Run tables and removes corresponding duplicated variable from PAOLINA/Summary table, as discussed in #671

mmkekic committed:
Change tests for modified Esmeralda

assert DST table inside file, assert summary table the same except
dropped columns, assert DST the same as in Penthesilea.

mmkekic committed:
Modify Esmeralda to pass tests

Adds kdst_from_df_writer, removes columns from summary table

mmkekic committed:
Add test eventinfo table contains all events

mmkekic committed:
Write event info and mc info without filtering in Esmeralda.

mmkekic committed:
Add test for load_event_summary in dst_functions.

mmkekic committed:
Add load_event_summary in dst_functions

mmkekic committed:
Remove norm_strat from Esmeralda; hard code to 'kr'

mmkekic committed:
Add check for no-filter MC events inside table in esmeralda_test.

mmkekic committed:
Remove time from PAOLINA/Summary in esmeralda

mmkekic committed:
Add time from Run/events to load_event_summary DataFrame output

mmkekic committed:
Cast to HitsCollection type only current event in esmeralda reader.

jahernando reviewed (approved):
This PR modifies esmeralda to output run_info and kdst without filters; reduces the common columns in summary and kdst, and provides the user a tool to extend the summary with variables from the kdst. It modifies the test accordingly.

The code is compact and clear, a little cryptic for those not so familiar with the capabilities of pandas. It will help to future code maintainers, to add some comments in plain english to translate what it is done in the code, sometimes, in one line, for example in the load_event_summary function.

If not other revisions of variables to be added or removed are needed in Esmeralda. I approve this PR.

mmkekic committed:
Add test to show bug in overlap_blob_energy in esmeralda_test

mmkekic committed:
Fix bug in overlap_blob_energy in esmeralda

mmkekic committed:
Add test to show repeated indices in load_dsts and a fix.

mmkekic committed:
Adapt track_blob_info_extractor for hits with nan energy

The output is track dataframe with nan hits removed, hits tables with
nan hits and a flag if the event contained nan hits.

mmkekic committed:
Add 'out_of_map' flag in sumamry table.

mmkekic committed:
Adapt Esmeralda body for changed filtering of events

The NN and paolina hits names are changed to low/high_th_hits. Filters
after corrections are one liners checking for length of
hits. Introduced new filter after track_blobs_extractor that checks if
there are any finite hits in event. Fixed output/input of
tracks_blob_extractor and event_sumamry_maker.

mmkekic committed:
Add 'out_of_map' flag assertation for test out_of_map_hits

mmkekic committed:
Add new test file and exact test comparison for esmeralda

mmkekic committed:
Change test file in test load_event_summary.

mmkekic committed:
Fix names in tests, adapts out_of_map test dataframe comparison

mmkekic committed:
Add comments in load_event_summary

gonzaponte commented on invisible_cities/reco/dst_functions.py:
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
+    print(kdst.columns)
+    if(len(kdst.s1_peak.unique()) != 1):
+        warnings.warn("Number of recorded S1 energies differs from 1 in event {}.Choosing first S1".format(event_number), UserWarning)
+    kdst_to_merge = kdst[['event', 'S1e', 'S1t', 'nS2', 'S2e', 'S2q']].groupby('event').agg({'S1e':lambda x:x.values[0],
```
`kdst.loc[:, columns]` is better (returns a view, not a copy)

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -277,27 +276,16 @@ def make_event_summary(event_number  : int              ,
+    es = pd.DataFrame(columns=['event', 'time', 'S2ec', 'S2qc', 'ntrks',
+                               'nhits', 'x_avg', 'y_avg', 'z_avg','r_avg',
+                               'x_min', 'y_min', 'z_min', 'r_min','x_max',
+                               'y_max', 'z_max', 'r_max'])
```
minor missalignment

gonzaponte commented on invisible_cities/reco/dst_functions_test.py:
```diff
@@ -86,3 +91,13 @@ def test_dst_event_id_selection_2():
+
+def test_load_event_summary_equal_to_old_summary(ICDATADIR):
+    old_esmeralda   =  os.path.join(ICDATADIR, "exact_Kr_tracks_with_MC.h5")
+    kdst_esmeralda =  os.path.join(ICDATADIR, "exact_Kr_tracks_with_MC_KDST.h5")
```
minor missalignment

gonzaponte commented on invisible_cities/reco/dst_functions.py:
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
+def load_event_summary(filename : str) -> pd.DataFrame :
+    summary = load_dst(filename, 'PAOLINA', 'Summary')
+    kdst    = load_dst(filename, 'DST'    , 'Events' )
+    print(kdst.columns)
```
forgotten print?

gonzaponte commented on invisible_cities/reco/dst_functions.py:
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
+    kdst    = load_dst(filename, 'DST'    , 'Events' )
+    print(kdst.columns)
+    if(len(kdst.s1_peak.unique()) != 1):
+        warnings.warn("Number of recorded S1 energies differs from 1 in event {}.Choosing first S1".format(event_number), UserWarning)
```
Do you have a protection agains nS1 = 0?

mmkekic replied:
No because it is always there, will give NaNs, the same as in input kdst table, guess its fine

gonzaponte commented on invisible_cities/reco/dst_functions.py:
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
+                                                                                             'S2e' : np.sum,
+                                                                                             'S2q' : np.sum}).reset_index()
+    kdst_to_merge.columns = ['event', 'S1e', 'S1t', 'nS2', 'S2e0', 'S2q0']
+    print(kdst_to_merge.columns)
```
another forgotten print?

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -6,6 +6,7 @@
 from hypothesis              import given
 from hypothesis              import settings
 from hypothesis.extra.pandas import columns, data_frames
```
it's not part of the PR, but please split this double import into two lines :)

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -416,7 +417,7 @@ def test_kdst_bins():
 
 
-@given(data_frames(columns=columns(kdst_variables, elements=floats(allow_nan=False))))
+@given(data_frames(index=range_indexes(min_size=1, max_size=10), columns=columns(kdst_variables, elements=floats(allow_nan=False))))
```
I think it would be good to define somewhere `kdst_dataframes = data_frames(...)` (or similar) for readability and to avoid repetition.

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -418,19 +417,14 @@ def esmeralda(files_in, file_out, compression, event_range, print_mod, run_numbe
 
-    threshold_and_correct_hits_paolina = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_paolina'], norm_strat = norm_strat),
+    threshold_and_correct_hits_paolina = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_paolina']),
                                                 args = 'hits',
```
Not a big deal, but the convention is to add spaces around equal signs in keyword arguments when each keyword is in a separate line, i.e.
`im_a_function(a, b, c=something, d=something_else)`
or
```
im_a_function(a, b,
              c = something,
              d = something_else)
```

Also, usually `**kwargs` go after hardcoded arguments. Like
`im_a_function(a, b, *c, d=something, **e)`

This happens in more places in the code, but I didn't mark them all. As I say is not a big deal, not worth the effort to look for them, but it is good to keep it in mind for future modifications to the code.

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -91,8 +91,6 @@ def threshold_and_correct_hits(hitc : evm.HitCollection) -> evm.HitCollection:
         t = hitc.time
         thr_hits = hif.threshold_hits(hitc.hits, threshold_charge     )
         mrg_hits = hif.merge_NN_hits ( thr_hits, same_peak = same_peak)
-        if len(mrg_hits) == 0:
-            return None
```
is this safe?

mmkekic replied:
yeah, the filter is checking if the len of hits is 0, the test  test_esmeralda_filters_events shows the events are properly filtered out

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -163,20 +145,16 @@ class HitEnergy(AutoNameEnumBase):
+        if out_of_map:
+            #add nan hits to track_hits, the track_id will be -1
+            track_hitc.hits.extend([h for h in hitc.hits if np.isnan   (getattr(h, energy_type.value))])
+            hits_without_nan       = [h for h in hitc.hits if np.isfinite(getattr(h, energy_type.value))]
```
alignment

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -186,63 +164,78 @@ def get_track_energy(track):
+                max_z = max([h.Z for v in t.nodes() for h in v.hits])
+                max_r = max([np.sqrt(h.X*h.X + h.Y*h.Y) for v in t.nodes() for h in v.hits])
+
+                pos = [h.pos for v in t.nodes() for h in v.hits]
```
I've seen this kind of thing in several places... This is a double loop repeated 7 times. I haven't time-profiled it, but my feeling is that it is very inefficient. I think it is worth to make a function that makes this loop only once and retrieves all the variables that may be needed. Opinions?

jacg replied:
> haven't time-profiled it, but my feeling is that it is very inefficient.

The most important rule of optimization: don't optimize anything until you've proven that it matters. That means that you must either prove (by timing) that 

a) this is a bottleneck
b) the alternative implementation is actually faster that the original.

If it's not a bottleneck, but the new version is faster and not significantly uglier, then go ahead. If the new version is uglier and more difficult to understand, then don't use it **unless profiling proves that the change makes a significant improvement in perfomance**.

However, the square brackets inside the `max` and `min` ensure that these will run in linear rather than constant space, which is just a pointless waste ... so get rid of the `[]`s. That is

    fn([ for ... in ... ])

should *always* be written as

    fn( for ... in ... )

or

   fn(( for ... in ... ))

if you like pointless characters (or there are other arguments besides the listcomp / genexpr).

jacg replied:
> Yes, I agree, but doesn't it grind your gears to see those loops repeated
so many times?

It does, it does. A smart enough compiler ...

mmkekic replied:
I agree! However I realized that we are already constructing pos list to use it in  np.average so I just rearranged the order to first construct pos list and then look for appropriate min/max

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -128,6 +128,10 @@ def test_esmeralda_with_out_of_map_hits(KrMC_hdst_filename_toy, correction_map_M
+    #assert event with nan energy labeled in summary_table
+    events_energy = df_hits_paolina.groupby('event')['Ec'].apply(pd.Series.sum, skipna=False)
+    np.testing.assert_array_equal( summary_table['out_of_map'], np.isnan(events_energy.values))
     #assert event number in EventInfo and MC/Extents iqual to nevt_req
```
`.Ec` instead of `['Ec']`?
`.out_of_map` instead of `['out_of_map']`?

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -284,3 +284,39 @@ def test_esmeralda_blob_overlap_bug(data_hdst, esmeralda_tracks, correction_map_
+              "Filters/low_th_select", "Filters/high_th_select", "Filters/topology_select"]
+
+    with tb.open_file(true_out)  as true_output_file:
+        with tb.open_file(file_out) as      output_file:
```
alignment

mmkekic committed:
Remove unnecesarry serial mark from tests.

mmkekic committed:
Remove option to run Esmeralda with uncorrected energy.

mmkekic committed:
Rename threshold parameters  in esmeralda.config

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
+            vox_size_z = voxels[0].size[2]
+            del(voxels)
+            def get_track_energy(track):
+                return sum([vox.Ehits for vox in track.nodes()])
```
this function here looks weird.. Isn't there a module for track functions?

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
+                length = plf.length(t)
+                numb_of_hits = len([h for vox in t.nodes() for h in vox.hits])
+                numb_of_voxels = len(t.nodes())
+                numb_of_tracks = len(tracks   )
```
alignment

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
+                blob_pos1, blob_pos2 = plf.blob_centres(t, blob_radius)
+
+                e_blob1, e_blob2, hits_blob1, hits_blob2 = plf.blob_energies_and_hits(t, blob_radius)
+                overlap = float(sum([h.Ec for h in set(hits_blob1).intersection(hits_blob2)]))
```
shouldn't it be `set(hits_blob1).intersection(set(hits_blob2))`?

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
+                                extr1_pos[0], extr1_pos[1], extr1_pos[2],
+                                extr2_pos[0], extr2_pos[1], extr2_pos[2],
+                                blob_pos1[0], blob_pos1[1], blob_pos1[2],
+                                blob_pos2[0], blob_pos2[1], blob_pos2[2],
```
`*extr1_pos, *extr2_pos, *blob_pos1, *blob_pos2` for readability?

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -98,29 +108,41 @@ def test_esmeralda_with_out_of_map_hits(KrMC_hdst_filename_toy, correction_map_M
+    summary_table       =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Summary')
+    #assert event with nan energy labeled in summary_table
+    events_energy = df_hits_paolina.groupby('event')['Ec'].apply(pd.Series.sum, skipna=False)
+    np.testing.assert_array_equal( summary_table['out_of_map'], np.isnan(events_energy.values))
```
Use `.column` instead of `[column]`?

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -151,11 +171,18 @@ def test_esmeralda_tracks_exact(data_hdst, esmeralda_tracks, correction_map_file
+    #make sure out_of_map is true for events not in df_tracks_exact
+    diff_events = list(set(df_tracks.event.unique()).difference(events))
+    df_summary = dio.load_dst(PATH_OUT, 'PAOLINA', 'Summary')
+    assert all(df_summary[df_summary.event.isin(diff_events)]['out_of_map'])
```
use `.loc`

gonzaponte reviewed (commented):
`test_esmeralda_exact_result_old` can be removed now or in a future PR. There is a commit in the history that proves the new esmeralda yields the same output and a new test to replace it.

mmkekic committed:
Add r_max and r_ave in tracks table

Modifly make_event_summary to use tracks table only (not loop over
paolina hits).

mmkekic committed:
Adapt test for different PAOLINA/Tracks table structure

mmkekic committed:
Remove MC and event_info check in test with out_of_map_hits.

Since esmeralda will not filter those events no point in checking they
are present in eventinfo and MC table. That check is done in
test_esmeralda_filters_events.
Fix alingment

mmkekic committed:
Cosmetics

mmkekic committed:
Restructure and rename  load_paolina_summary.

The function can take as input either a single esmeralda output or a
list of outputs.

mmkekic committed:
Add test for list of files as input for load_paolina_summary.

mmkekic committed:
Replace file for Esmeralda exact table comparison

mmkekic committed:
Add get_track_energy to paolina functions and corresponding test

mmkekic committed:
Use get_track_energy from paolina_functions in Esmeralda

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -142,12 +122,6 @@ def track_blob_info_creator_extractor(vox_size         : [float, float, float],
     ----------
     vox_size         : [float, float, float]
         (maximum) size of voxels for track reconstruction
-    energy_type      : HitEnergy
-        class HitEnergy(AutoNameEnumBase):
-            E        = auto()
-            Ec       = auto()
-        energy attribute to use for voxelization/ tracks
-
```
Is `HitEnergy` still used somewhere else? If not remove it.

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -164,96 +138,102 @@ class HitEnergy(AutoNameEnumBase):
+                min_y = min(pos, key = lambda x : x[1])[1]
+                max_y = max(pos, key = lambda x : x[1])[1]
+                min_z = min(pos, key = lambda x : x[2])[2]
+                max_z = max(pos, key = lambda x : x[2])[2]
```
mmm, I was thinking about something in the lines of
```
pos = bla
x, y, z = map(np.array, zip(*pos))
r = np.sqrt(x**2 + y**2)
```
and then mins and maxes are trivial.

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -164,96 +138,102 @@ class HitEnergy(AutoNameEnumBase):
+                try:
+                    types_dict
+                except NameError:
+                    types_dict = dict(zip(df.columns, [type(x) for x in list_of_vars]))
```
it bothers me a little bit that this thing that is actually constant is being defined here in the middle of everything. But I am not sure the alternative is better. I would try to define the df columns in the top of the file together with this type mapping. BTW, why is it needed?

gonzaponte replied:
@mmkekic digit: "if i dont change datatype it will be  floats and we dont want that"

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -277,52 +257,44 @@ def make_event_summary(event_number  : int              ,
+    try:
+        S2qc  = max(-1, sum(topology_info.charge))
+    except AttributeError:
+        S2qc  = -1
```
`S2qc = max(-1, sum(topology_info.charge)) if hasattr(topology_info, "charge") else -1`?

gonzaponte commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -277,52 +257,44 @@ def make_event_summary(event_number  : int              ,
         x_avg /= S2ec
         y_avg /= S2ec
         z_avg /= S2ec
         r_avg /= S2ec
```
isn't `np.average` better?

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -60,34 +62,42 @@ def test_esmeralda_filters_events(KrMC_hdst_filename_toy, correction_map_MC_file
+    with tb.open_file(PATH_OUT)  as h5out:
+        MC_num_evs = h5out.root.MC.extents[:]['evt_number']
+        assert len(MC_num_evs) == nevt_req
+
```
why open it twice?

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -166,31 +184,56 @@ def test_esmeralda_exact_result(ICDATADIR, KrMC_hdst_filename, correction_map_MC
+    tables_in = ( "MC/extents"  , "MC/hits"       , "MC/particles"  , "MC/generators",
+                  "DST/Events"  , "Run/events"    , "Run/runInfo"                    )
+    with tb.open_file(file_in)  as true_output_file:
+        with tb.open_file(file_out) as      output_file:
```
minor missalignment

gonzaponte commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -200,7 +243,71 @@ def test_esmeralda_empty_input_file(config_tmpdir, ICDATADIR):
+                         blob_radius              = 21 * units.mm)        ))
+    cnt = esmeralda(**conf)
+
+    df_tracks           =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Tracks')
```
extra spaces that align with nothing

gonzaponte reviewed (commented):
minor things

mmkekic committed:
Prettyfy esmeralda a la Gonzalo.

mmkekic committed:
Prettify esmeralda_test a la Gonzalo

gonzaponte reviewed (approved):
Awesome job by @mmkekic to include the new summary structure into esmeralda. Approved!
